File: src/code/javascript/graph/kahn.js
const kahnTopologicalSort = (graph) => {
    const result = []
    const queue = []
    const indegree = new Map()

    for (let nodes of Object.values(graph)) {
        for (let node of nodes) {
            if (indegree.has(node)) {
                indegree.set(node, indegree.get(node) + 1)
            } else {
                indegree.set(node, 1)
            }
        }
    }

    for (let node of Object.keys(graph)) {
        if (!indegree.has(node)) {
            queue.push(node)
        }
    }

    while (queue.length > 0) {
        const node = queue.shift()
        result.push(node)

        if (graph[node]) {
            for (let neighbor of graph[node]) {
                indegree.set(neighbor, indegree.get(neighbor) - 1)

                if (indegree.get(neighbor) === 0) {
                    queue.push(neighbor)
                }
            }
        }
    }

    return result.length === Object.keys(graph).length ? result : []
}

File: src/code/ruby/graph/kahn.rb
def kahn_topological_sort(graph)
  result    = []
  indegree  = Hash.new(0)

  graph.each_value do |vertices|
    vertices.each do |v|
      indegree[v] += 1
    end
  end

  queue = graph.keys.select { |node| indegree[node] == 0 }

  until queue.empty?
    node = queue.shift
    result << node

    graph[node].each do |neighbor|
      indegree[neighbor] -= 1
      queue << neighbor if indegree[neighbor] == 0
    end
  end

  result.size == graph.size ? result : []
end
